### PR TITLE
[IN-112]: add deployment step

### PR DESCRIPTION
**Pull request by electather**
# Description

## What is the motivation for this pull request?

This PR adds required configurations for Circle CI to run test, build and deploy for the plugin. 

## What is the solution?

N/A

## What areas of the site does it impact?

Build and deploy

## How to test?

- [ ] Build and deploy 

## Other Notes

N/A

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -67,50 +96,117 @@ commands:
...
+    environment:
+      BASE_URL: https://api.london.dev.w3w.io/v3
+      PACKAGE_NAME: what3words-dev
     steps:
       - checkout
       - setup_environment
       - node/install-packages
       - run: npm run coverage
-      - run: npm run compile
+      - run: npm run build
```
Comment (c5haw) [inline .circleci/config.yml:116]:
```suggestion
      - run: npm run compile
```
Accidental change?

**Review by c5haw — COMMENTED** [review]
Are there some missing files from the commit?

File: .circleci/config.yml (hunk 1 above)
Comment (electather) [reply .circleci/config.yml:116]: no, in this case, [we need the `build` script to be run](https://github.com/w3w-internal/bigcommerce-storefront-plugin/blob/1e4aa254fcfdab7f67678ecf4d2ffcbe9d17449d/package.json#L29).
This PR builds upon #1  but it's not merged yet.

File: .circleci/config.yml [hunk 2]
```diff
@@ -67,50 +96,117 @@ commands:
...
+            - ./build
 
-  publish:
+  build-prod:
+    executor: build-executor
+    environment:
+      PACKAGE_NAME: what3words
+    steps:
+      - checkout
+      - setup_environment
```
Comment (c5haw) [inline .circleci/config.yml:130]:
```suggestion
```

File: .circleci/config.yml [hunk 3]
```diff
@@ -67,50 +96,117 @@ commands:
...
+          name: 'install basic utility tools'
+          command: yum install -y tar gzip which git
+      - attach_workspace:
+          at: /tmp/workspace
+      - setup_environment:
+          version_path: /tmp/workspace/
+      - aws-s3/sync:
+          from: /tmp/workspace/build/
+          to: 's3://<< pipeline.parameters.dev-s3-bucket >>/bigcommerce-storefront-plugin-preprod/'
+          arguments: --delete
```
Comment (c5haw) [inline .circleci/config.yml:188]: Remove this block as its redundant.

File: .circleci/config.yml [hunk 4]
```diff
@@ -67,50 +96,117 @@ commands:
...
+          command: yum install -y tar gzip which git
+      - attach_workspace:
+          at: /tmp/workspace
+      - setup_environment:
+          version_path: /tmp/workspace/
+      - aws-s3/sync:
+          from: /tmp/workspace/build/
+          to: 's3://<< pipeline.parameters.prod-s3-bucket >>/bigcommerce-storefront-plugin/'
+          arguments: --delete
   tag:
```
Comment (c5haw) [inline .circleci/config.yml:204]: Remove this job

**Review by c5haw — CHANGES_REQUESTED** [review]
I like the gitops approach of using Github Releases (and tags) to trigger deployments. This aligns with what I was envisioning for triggering publishing libraries/plugins. If we are going to go down this approach though I think theres a few things think you should do here some of which should be discussed and potentially replicated into the library boilerplate/libraries elsewhere:

1. Use semantic versioning when we create a release/tag
2. Following semver we should use the [prelease semantic](https://github.com/npm/node-semver#functions) as the trigger for dev CDN so `/^(?:[0-9]+\.){2}\.[0-9]+-alpha\.[0-9]+$/` (and without prerelease tag for production)

I have an example of this [here](https://github.com/w3w-internal/javascript-sdk/pull/60/files#diff-78a8a19706dbd2a4425dd72bdab0502ed7a2cef16365ab7030a5a0588927bf47R179-R273) that triggers publishing prerelease or latest depending on the version regex test using circleci `when...unless` conditionals. Might inspire something here (and other projects/boilerplate)
3. Setting the version env var to use the tag first or fallback to the version in `package.json`
  `echo "export VERSION=${CIRCLE_TAG:-$(cat << parameters.version_path >>/package.json | grep '\"version\":' | sed -e 's/ \"version\": \"//' | sed -e 's/\",//')}" >> $BASH_ENV`
4. Naming convention; for libraries I tend to use the terminology publish rather than deploy
5. Remove preprod CDN as this doesn't exist and just replicates dev

File: .circleci/config.yml (hunk 1 above)
Comment (electather) [reply .circleci/config.yml:116]: the main is (was) just boilerplate, and #1 had nothing to do with the deployment step. this was done to reduce the dependency of the PRs on each other, so they can be merged independently.

File: .circleci/config.yml (hunk 4 above)
Comment (electather) [reply .circleci/config.yml:204]: This is the deployment step for prod. do you have any replacements in mind?

**Comment by electather** [comment]
@c5haw I updated the deployment with my spin on the regex. have a look.

File: .circleci/config.yml (hunk 4 above)
Comment (c5haw) [reply .circleci/config.yml:162]: It highlighted the wrong line. This was meant to be highlighting the `tag` job

File: .circleci/config.yml [hunk 5]
```diff
@@ -1,5 +1,25 @@
...
+    type: string
+    default: 'cdn.dev.w3w.io'
+  prod-s3-bucket:
+    type: string
+    default: 'cdn.what3words.com'
+
+only-prerelease-tags: &only-prerelease-tags
+  filters:
+    tags:
+      only: /^v([0-9]+)\.([0-9]+)\.([0-9]+)(?:-([0-9A-Za-z-]+(?:\.[0-9]+)*))?(?:\+[0-9]+)?$/
```
Comment (c5haw) [inline .circleci/config.yml:14]:
```suggestion
      only: /^v([0-9]+)\.([0-9]+)\.([0-9]+)(?:-([0-9A-Za-z-]+(?:\.[0-9]+)*)){1}(?:\+[0-9]+)?$/
```
If this is only prerelease then I presume it should have the prerelease suffix

File: .circleci/config.yml [hunk 6]
```diff
@@ -67,50 +89,82 @@ commands:
...
 jobs:
-  build_test:
+  build-dev:
     executor: build-executor
+    environment:
+      BASE_URL: https://api.london.dev.w3w.io/v3
+      PACKAGE_NAME: what3words-dev
     steps:
       - checkout
       - setup_environment
```
Comment (c5haw) [inline .circleci/config.yml:106]:
```suggestion
```
Can remove this job as the env vars that are set there are not being used anywhere.

File: .circleci/config.yml [hunk 7]
```diff
@@ -67,50 +89,82 @@ commands:
...
+          command: yum install -y tar gzip which git
       - attach_workspace:
           at: /tmp/workspace
       - setup_environment:
-          version_path: /tmp/workspace
-      - run:
-          name: "publish npm"
-          command: cd /tmp/workspace && npm publish
-      # - jobstatus
+          version_path: /tmp/workspace/
```
Comment (c5haw) [inline .circleci/config.yml:142]:
```suggestion
```
Can remove this job as the env vars that are set there are not being used anywhere.

File: .circleci/config.yml [hunk 8]
```diff
@@ -67,50 +89,82 @@ commands:
...
+  publish:
+    executor: publish-executor
+    steps:
+      - run:
+          name: 'install basic utility tools'
+          command: yum install -y tar gzip which git
+      - attach_workspace:
+          at: /tmp/workspace
+      - setup_environment:
+          version_path: /tmp/workspace/
```
Comment (c5haw) [inline .circleci/config.yml:157]:
```suggestion
```
Can remove this job as the env vars that are set there are not being used anywhere.

**Review by c5haw — CHANGES_REQUESTED** [review]
One other question. The way this writes to s3 means that you can only ever have a single version published at any time. Do we want to be able to point to a specific version at any point? Something like:

```
        aws-s3/sync:
          from: /tmp/workspace/build/
          to: 's3://<< pipeline.parameters.dev-s3-bucket >>/bigcommerce-storefront-plugin@<< parameters.version >>/'
          arguments: --delete
```

And pass in the version taken from the `VERSION` env var set in `setup_environment`

**Review by c5haw — APPROVED** [review]
Still one outstanding question but it'll come up again at a later date and we can address then - LGTM

**Comment by electather** [comment]
@c5haw I didn't see this question:
> One other question. The way this writes to s3 means that you can only ever have a single version published at any time. Do we want to be able to point to a specific version at any point? Something like:

having multiple versions doesn't really do anything for us, in case of ecwid, the process of changing version is time consuming by ecwid, and it's done once for everyone, users can't really pick their versions, it's always gonna be latest. 
as for development / beta application, we have that, in a form of dev S3 bucket. 
as for bigcommerce, we can deploy different versions, but then managing deployed scripts and bumping script versions for our users ( per user basis ) is gonna be a challange. 
so in my mind, best way to go is with reproducable buids, just running an older tag's job in circle CI will overwrite the exisiting versions. so we don't have to keep diffrent versions on our bucket.

**Pull request merged**